src/gnc_attitude_estimation.cpp

```diff
@@ -0,0 +1,123 @@
+//
+// src/gnc_attitude_estimationr.cpp
+// PSim
+//
+// Contributors:
+//   Kyle Krol  kpk63@cornell.edu
+//
+// Pathfinder for Autonomous Navigation
+// Space Systems Design Studio
+// Cornell Univeristy
+//
+
+#include <gnc_attitude_estimation.hpp>
+#include <gnc_environment.hpp>
+#include <gnc_utilities.hpp>
+
+#include <cmath>
+
+static_assert(std::numeric_limits<double>::has_quiet_NaN,
+    "GNC code requires quiet NaN's to be available.");
+static_assert(std::numeric_limits<float>::has_quiet_NaN,
+    "GNC code requires quiet NaN's to be available.");
+
+namespace gnc {
+
+AttitudeEstimatorState::AttitudeEstimatorState() {
+  this->t = std::numeric_limits<double>::quiet_NaN();
+  this->q_body_eci = lin::nans<decltype(this->q_body_eci)>();
+  this->w_buffer = CircularBuffer<lin::Vector3f, 5>();
+}
+
+AttitudeEstimatorData::AttitudeEstimatorData() {
+  this->t = std::numeric_limits<double>::quiet_NaN();
+  this->r_ecef = lin::nans<decltype(this->r_ecef)>();
+  this->b_body = lin::nans<decltype(this->b_body)>();
+  this->s_body = lin::nans<decltype(this->s_body)>();
+}
+
+AttitudeEstimate::AttitudeEstimate() {
+  this->q_body_eci = lin::nans<decltype(this->q_body_eci)>();
+  this->w_body = lin::nans<decltype(this->w_body)>();
+}
+
+/** @fn estimate_q_body_eci
+ *  @returns Zero of success.
+ *  Attempts to determine attitude given time, position, a sun vector, and
+ *  magnetic field vector. The result, if succesful, is written into the
+ *  quaternion q_body_eci. */
+static int estimate_q_body_eci(double t, lin::Vector3d const &r_ecef,
+    lin::Vector3f const &s_body, lin::Vector3f const &b_body, lin::Vector4f &q_body_eci) {
+  // Setup the rest of the triad inputs (including expected sun and magnetic
+  // field vectors in ECI)
+  lin::Vector3f s_eci, b_eci, r_ecef_f({
+    static_cast<float>(r_ecef(0)),
+    static_cast<float>(r_ecef(1)),
+    static_cast<float>(r_ecef(2))
+  });
+  env::sun_vector(t, s_eci);
+  env::magnetic_field(t, r_ecef_f, b_eci);
+  // Perform triad and return the result code (result is written to q_body_eci)
+  return utl::triad(s_eci, b_eci, s_body, b_body, q_body_eci);
+}
+
+/** @fn estimate_w_body
+ *  Approximates the angular rate of a body given a quaternion q_m measured at a
+ *  time t and a quaternion q_n measured at a time t + dt. */
+static void estimate_w_body(double dt, lin::Vector4f const &q_m,
+    lin::Vector4f const &q_n, lin::Vector3f &w) {
+  // Find the 'difference' between the two quaternions
+  lin::Vector4f q_n_conj, dq;
+  utl::quat_conj(q_n, q_n_conj);
+  utl::quat_cross_mult(q_m, q_n_conj, dq);
+  // Approximate w_body
+  w = lin::ref<3, 1>((2.0f / static_cast<float>(dt)) * dq, 0, 0);
```

Comment:
Check the sign of `dq[4]`, if it is negative, negate `w` as well. This is because this calc assumes `dq[4]` is positive, so if it isn't just negate all of dq.